lib/worker.js

```diff
@@ -0,0 +1,211 @@
+    payload = JSONParse(payload);
+    switch (messageType) {
+      case WorkerContextBinding.USER:
+        return workerObject.emit('message', payload);
+      case WorkerContextBinding.INTERNAL:
+        assert.fail('unreachable');
+      case WorkerContextBinding.ERROR:
+        return onerror(payload);
+      default:
+        assert.fail('unreachable');
+    }
+  };
+};
+
+Worker.prototype._checkAlive = function() {
+  if (!this._workerContext)
+    throw new RangeError('this worker has been terminated');
+};
+
+Worker.prototype._doPostMessage = function(payload, transferList, type) {
+  this._checkAlive();
+
+  if (transferList == null)
+    transferList = EMPTY_ARRAY;
+  else if (!Array.isArray(transferList))
+    throw new Error('transferList must be an array');
+
+  this._workerContext.postMessage(JSONStringify(payload), transferList, type);
+};
+
+Worker.prototype.postMessage = function(payload, callback) {
+  this._doPostMessage(payload, null, WorkerContextBinding.USER);
+};
+
+Worker.prototype.terminate = function(callback) {
+  this._checkAlive();
+  var context = this._workerContext;
+  this._workerContext = null;
+  if (typeof callback === 'function') {
+    this.on('exit', function(exitCode) {
+      callback(null, exitCode);
+    });
+  }
+  context.terminate();
+};
+
+Worker.prototype.ref = function() {
+  this._checkAlive();
+  this._workerContext.ref();
+};
+
+Worker.prototype.unref = function() {
+  this._checkAlive();
+  this._workerContext.unref();
+};
+
+if (process.isWorkerInstance) {
+  const postMessage = function(payload, transferList, type) {
+    if (transferList == null)
+      transferList = EMPTY_ARRAY;
+    else if (!Array.isArray(transferList))
+      throw new Error('transferList must be an array');
+
+    WorkerContextBinding.workerWrapper._postMessage(JSONStringify(payload),
+                                                    transferList,
+                                                    type);
+  };
+  const workerFatalError = function(er) {
+    var defaultStack = null;
+    var defaultMessage = '[to string conversion failed]';
+    var defaultBuiltinType = 'Error';
+
+    var message = defaultMessage;
+    var builtinType = defaultBuiltinType;
+    var stack = defaultStack;
+    var additionalProperties = {};
+
+    if (er instanceof Error) {
+      try {
+        builtinType = er.name;
+        if (typeof builtinType !== 'string')
+          builtinType = defaultBuiltinType;
+      } catch (ignore) {}
+
+      try {
+        stack = er.stack;
+        if (typeof stack !== 'string')
+          stack = defaultStack;
+      } catch (ignore) {}
+      try {
+        // Get inherited enumerable properties.
+        // .name, .stack and .message are all non-enumerable
+        for (var key in er)
+          additionalProperties[key] = er[key];
+        // The main message delivery must always succeed.
+        JSONStringify(additionalProperties);
```

Comment:
This is called from fatalError handler, if it throws, then the real cause for the fatal error will not be discovered